addons/udes_security/controllers/main.py

```diff
@@ -5,12 +5,100 @@
+        `_sequence` parameter, so there is no risk of colliding ids from separate models.
+
+        They are stored in the db like "ir.actions.act_window,90" or "ir.actions.server,242"
+        due to being a Reference field, so perform a fuzzy search on %,<id>% to get it
+
+        Be extra safe and ensure `action_id` is an int, as if the user enters text in the URL
+        for action= param, it will come through to here as a string
+
+        :param: action_id: int() for an action (ir.actions.server, ir.actions.act_window, etc..)
+
+        :return: ir.ui.menu(x,) recordset, will be empty if a non existent action_id is passed in
+        """
+        IrUiMenu = request.env["ir.ui.menu"]
+
+        if not isinstance(action_id, int):
+            raise ValidationError("Invalid action ID specified.")
+        # Escape the leading and trailing static % symbols with %,
+        # as we don't want to string format them in the execute() call
+        query = "SELECT id FROM ir_ui_menu WHERE action LIKE '%%,%s%%'"
+        request.env.cr.execute(query, (action_id,))
+        result = request.env.cr.fetchone()
+        return IrUiMenu.sudo().browse(result)
+
+    def _user_has_access_right(self, action, menu, user):
+        """
+        Check user access rights against action, menu, and menu parent.
+
+        Start at the provided action & menu, and recursively call, passing in the menu parent until either:
+        1.) The user has no groups which are specified against the action or menu (return False)
+        2.) The user has groups specified against the action or menu, and no parent exists (return True)
+          2.2) There are no groups specified against the action or menu, and no parent exists (return True)
+
+        This ensures that the user has access at least every menuitem in the chain of menuitems
+        leading to the menu/action they are trying to access, following the same logic of the web UI.
+
+        :param: action: ir.actions.act_window(x,) recordset
+        :param: menu: ir.ui.menu(x,) recordset
+        :param: user: res.users(x,) recordset
+
+        :return: bool() whether or not the user should be granted access
+        """
+        # TODO(palabaster): This line makes up 25% of this entire functions cpu time, any way we can improve?
+        parent = menu.parent_id
+        group_ids_for_access = []
+        # ir.actions.client, ir.actions.act_url don't have groups_id parameter
+        if hasattr(action, "groups_id"):
+            group_ids_for_access += action.groups_id.ids
+        # Can have actions defined in the system with no menu associated. These should be blocked.
+        if not menu:
+            return False
+        group_ids_for_access += menu.groups_id.ids
+        user_has_any_groups = any(g_id in user.groups_id.ids for g_id in group_ids_for_access)
+        if (group_ids_for_access and user_has_any_groups) or not group_ids_for_access:
```

Comment:
Correct me if i am wrong.
In case that menu has groups `group_ids_for_access` is a list with some ids and user doesnt have any of this groups i think should return false not search to the parent menu isnt it?

Reply:
if a menu/action has groups, and the user does not have any of those groups, it will hit the `else` block and return `False` (no need to continue searching upward)

Reply:
```
         if group_ids_for_access and not user_has_any_groups:
             return False
        else:
            return self._user_has_access_right(action, parent, user) if parent else True
```
How do you think to change the way of condition seems little confusing the first one

Reply:
yep that works, nice to null out the or condition in the first block, didn't see that one :)

Reply:
just for ref, this is what i originally had the function as.. not confusing at all
```python
        parent = menu.parent_id
        if action and action.groups_id:
            action_access = any(g in user.groups_id.ids for g in action.groups_id.ids)
            if action_access:
                if parent:
                    return self._user_has_access_right(action, parent, user)
                else:
                    return True
            else:
                return False
        if menu and menu.groups_id:
            menu_access = any(g in user.groups_id.ids for g in menu.groups_id.ids)
            if menu_access:
                if parent:
                    return self._user_has_access_right(action, parent, user)
                else:
                    return True
            else:
                return False
        if menu and parent:
            return self._user_has_access_right(action, parent, user)

        return True
```